client/service/src/chain_ops.rs

```diff
@@ -114,86 +218,100 @@ impl<
 		// This makes it possible either to interleave other operations in-between the block imports,
 		// or to stop the operation completely.
 		let import = future::poll_fn(move |cx| {
-			// Start by reading the number of blocks if not done so already.
-			let count = match count {
-				Some(c) => c,
-				None => {
-					let c: u64 = match Decode::decode(&mut io_reader_input) {
-						Ok(c) => c,
-						Err(err) => {
-							let err = format!("Error reading file: {}", err);
-							return std::task::Poll::Ready(Err(From::from(err)));
-						},
-					};
-					info!("📦 Importing {} blocks", c);
-					count = Some(c);
-					c
-				}
-			};
-
 			// Read blocks from the input.
-			if read_block_count < count {
-				match SignedBlock::<Self::Block>::decode(&mut io_reader_input) {
-					Ok(signed) => {
-						let (header, extrinsics) = signed.block.deconstruct();
-						let hash = header.hash();
-						// import queue handles verification and importing it into the client
-						queue.import_blocks(BlockOrigin::File, vec![
-							IncomingBlock::<Self::Block> {
-								hash,
-								header: Some(header),
-								body: Some(extrinsics),
-								justification: signed.justification,
-								origin: None,
-								allow_missing_state: false,
-								import_existing: force,
-							}
-						]);
-					}
-					Err(e) => {
-						warn!("Error reading block data at {}: {}", read_block_count, e);
-						return std::task::Poll::Ready(Ok(()));
+			match Pin::new(&mut block_stream).poll_next(cx) {
+				Poll::Ready(None) => {
+					let read_block_count = block_stream.read_block_count();
+					if let Some(count) = block_stream.count() {
```

Comment:
Why do we print once `count` and the other time `read_block_count`?

Reply:
Because `BlockStream::Json` does not have a `count` field (same issue [here](https://github.com/paritytech/substrate/pull/5816#discussion_r416793180)).
When we export to binary, we first write the number of blocks, and then the blocks. So when we decode from binary, we know exactly how many blocks we are expecting to decode (`count`).

However when we export to json, we simply write all the blocks one after another. So we do not have access to the `count` varbiable.

I would be very keen to hear an elegant way of solving this problem! :)

Reply:
This does not makes any sense here. `read_block_count == count` at this position and if not, than there is a bug.

Reply:
Yes both should be equal indeed, that is why we can print one or the other. I'll print `read_block_count` then! :)

Reply:
Yep